Packet++/src/TcpLayer.cpp

```diff
@@ -381,6 +382,8 @@ void TcpLayer::parseNextLayer()
 		m_NextLayer = new FtpResponseLayer(payload, payloadLen, this, m_Packet);
 	else if (FtpLayer::isFtpPort(portDst) && FtpLayer::isDataValid(payload, payloadLen))
 		m_NextLayer = new FtpRequestLayer(payload, payloadLen, this, m_Packet);
+	else if (SomeIpLayer::isSomeIpPort(portSrc) || SomeIpLayer::isSomeIpPort(portDst) || SomeIpLayer::isDataValid(payload, payloadLen))

```

Comment:
This if statement means that every payload which meets the `SomeIpLayer::isDataValid()` criteria will be considered as SOME/IP, regardless of SOME/IP ports. Is this the intended behavior?

I think a safer approach could be:
```cpp
else if ((SomeIpLayer::isSomeIpPort(portSrc) || SomeIpLayer::isSomeIpPort(portDst)) && SomeIpLayer::isDataValid(payload, payloadLen))
{
    ...
}
```

Reply:
not sure about this, currently there is a problem with the `SomeIpLayer::addSomeIpPort(...)` method which produces a memory leak if it is called (uses a `std::unordered_set<uint16_t>`). We could add a method for deallocation but i think it is very annoying if this is not done automatically. Do you have a suggestion?

Reply:
@kolbex I have a suggestion: we can go with an approach that combines Pcap++Test's [`testSetUp`](https://github.com/seladb/PcapPlusPlus/blob/4708cfbeb901551c013fbad0da13600233e254b7/Tests/Pcap%2B%2BTest/Common/TestUtils.cpp#L94) and [`DeviceTeardown`](https://github.com/seladb/PcapPlusPlus/blob/4708cfbeb901551c013fbad0da13600233e254b7/Tests/Pcap%2B%2BTest/Common/TestUtils.h#L9):

1. We can add `pcpp::SomeIpLayer::removeSomeIpPort(uint16_t port)` so we can remove a port from the static `unordered_set`
2. Then we can add a new `testSetUp()` method in [`Packet++Test/Utils/TestUtils.h`](https://github.com/seladb/PcapPlusPlus/blob/master/Tests/Packet%2B%2BTest/Utils/TestUtils.h) and [`Packet++Test/Utils/TestUtils.cpp`](https://github.com/seladb/PcapPlusPlus/blob/master/Tests/Packet%2B%2BTest/Utils/TestUtils.cpp) that will add and remove a dummy port, just to initialize the `unordered_set`:

   ```cpp
   void testSetUp()
   {
   	pcpp::SomeIpLayer::addSomeIpPort(1);
   	pcpp::SomeIpLayer::removeSomeIpPort(1);
   }
   ```

3. `testSetUp()` should run in `Packet++Test/main.cpp` just about [here](https://github.com/seladb/PcapPlusPlus/blob/4708cfbeb901551c013fbad0da13600233e254b7/Tests/Packet%2B%2BTest/main.cpp#L109) before the tests start running
4. Now we can add a simple class `SomeIpTeardown` in `SomeIpTests.cpp` that will add a port in its c'tor and remove it in its d'tor, like this:

   ```cpp
   class SomeIpTeardown
   {
   private:
   	uint16_t m_SomeIpPort;
   public:
   	SomeIpTeardown(uint16_t someIpPort)
   	{
   		m_SomeIpPort = someIpPort;
   		pcpp::SomeIpLayer::addSomeIpPort(someIpPort);
   	}
   
   	~SomeIpTeardown()
   	{
   		pcpp::SomeIpLayer::removeSomeIpPort(m_SomeIpPort);
   	}
   };
   ``` 
5. The last step is to create an instance of `SomeIpTeardown` in each test, like this:

   ```cpp
   PTF_TEST_CASE(SomeIpParsingTest)
   {
   	timeval time;
   	gettimeofday(&time, NULL);
   
   	SomeIpTeardown(1111);
   
   	...
   ```

   What this does is ensure that a port is added in the beginning of the test and remove when the test ends or fails (which is when the `SomeIpTeardown` d'tor is called


I know these are quite a lot of changes, but I think it's a pretty clean solution to the memleak problem. Please let me know what you think

Reply:
Looks nice, two Questions:

> What is the `testSetUp()` method for?

> I would also add a method `pcpp::SomeIpLayer::removeAllSomeIpPorts(uint16_t port)` which removes all entries of the unordered_set. In order when more ports are added that you do not have to remember each port. What do you think?

Reply:
> Looks nice, two Questions:
> 
>     1. What is the `testSetUp()` method for?
> 

It's just to initialize the `unordered_set` in the first time, otherwise you'll see "leaked memory" for the object initialization (even if the `unordered_set` is empty)

>     2. I would also add a method `pcpp::SomeIpLayer::removeAllSomeIpPorts(uint16_t port)` which removes all entries of the unordered_set. In order when more ports are added that you do not have to remember each port. What do you think?

Do you mean  `pcpp::SomeIpLayer::removeAllSomeIpPorts()` without the `port` parameter? 🤔

Reply:
Yes, without the port:
```
void SomeIpLayer::removeAllSomeIpPorts()
{
	m_SomeIpPorts.clear();
}
```

Reply:
Sounds good!

Reply:
I currently updated the PR, i did not use the `SomeIpTeardown` class because i wanted to test both the `removeAllSomeIpPorts` and `removeSomeIpPort` methods, please take a look, i hope we can merge this PR.

Reply:
Sure, I think we can merge the PR as is. I suggest we still add the `SomeIpTeardown` class in a separate PR to make sure we remove all SOME/IP ports after the test. It can have a simpler implementation:

```cpp
class SomeIpTeardown
{
public:
	SomeIpTeardown() {}

	~SomeIpTeardown()
	{
		pcpp::SomeIpLayer::removeAllSomeIpPorts();
	}
};
```